safehttp/plugins/xsrf/xsrf.go

```diff
@@ -22,142 +22,234 @@ import (
 	"crypto/rand"
 	"encoding/base64"
 	"errors"
-	"fmt"
-
 	"github.com/google/go-safeweb/safehttp"
 	"golang.org/x/net/xsrftoken"
 )
 
 const (
-	// TokenKey is the form key used when sending the token as part of POST
-	// request.
-	TokenKey    = "xsrf-token"
 	cookieIDKey = "xsrf-cookie"
+	TokenKey    = "xsrf-token"
 )
 
+type tokenCtxKey struct{}
+
 var statePreservingMethods = map[string]bool{
 	safehttp.MethodGet:     true,
 	safehttp.MethodHead:    true,
 	safehttp.MethodOptions: true,
 }
 
-// Interceptor implements XSRF protection.
 type Interceptor struct {
-	// SecretAppKey uniquely identifies each registered service and should have
-	// high entropy as it is used for generating the XSRF token.
-	SecretAppKey string
+	secretAppKey string
+	g            Generator
+	c            Checker
+	i            Injector
```

Comment:
`Checker` should check whether the request follows XSRF protection protocol (i.e. for state-changing it should require an XSRF token; the `Checker` is responsible for taking the request and determining whether it is fine or not). If not, I'd make the `Checker` return that information as opposed to directly writing the response. The XSRF plugin itself should do it.

`Generator` should generate the XSRF tokens (or whatever needs to be in a cookie). It's not clear to me what it should return, but I don't think it should do anything with the `ResponseWriter`.

`Injector` should get whatever `Generator` returned and apply it to the safe response. Due to the fact that cookies are tied to `ResponseWriter` and not to the safe response itself, it needs to accept a `ResponseWriter` alongside the safe response.

cc: @empijei to double-check we have a shared vision.

Reply:
> Checker should check whether the request follows XSRF protection protocol (i.e. for state-changing it should require an XSRF token; the Checker is responsible for taking the request and determining whether it is fine or not). If not, I'd make the Checker return that information as opposed to directly writing the response. The XSRF plugin itself should do it.

I agree

> Generator should generate the XSRF tokens (or whatever needs to be in a cookie). It's not clear to me what it should return, but I don't think it should do anything with the ResponseWriter.

We cannot generalize this: some defences will be a cookie, some others will be a token for the template, some will be headers. How would you do it? And what would you inject, where and how?

Reply:
The `Injector` should probably take care of ensuring all protection is there (i.e. it should set cookies and/or inject the template with the token, based on the case) so "inject the protection". Currently, generating the token is also done in the `Injector`. It is rather hard to abstract away the `Generator` (the thing that should generate the token) from the `Injector`, as the info it needs depends on whether the plugin provides protection for Angular or for the "default" case as well as what crypto stuff it uses to generate the token. Hence, a `Generate` function would take different arguments based on the case. One alternative I could think of is the following:
```
type Generator interface {
Generate(r *safehttp.IncomingRequest, data ...interface{}) (string, error)
}
```
Then, in the "default case":
- we pass the `secretAppKey` and `cookieID` in the `Inject` function alongside the request and `Generate` generates the token either using Tink or `xsrftoken`

and in the Angular case:
- we could pass a `secretAppKey with the request and the token to be set in the cookie could be generated using that key or simply using `crypto/rand` (afaiu, pre-login protection is not an issue in Angular, right?)

Reply:
As discussed: focus on the interfaces that `xsrf` plugin needs first.

One good exercise you can do: write comments for your interface methods. It will make it clearer to you how to divide the responsibilities.